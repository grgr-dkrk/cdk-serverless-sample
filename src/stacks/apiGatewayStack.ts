import { Construct, Stack, StackProps } from '@aws-cdk/core'
import { Function as LambdaFunction } from '@aws-cdk/aws-lambda'
import { LambdaIntegration, RestApi } from '@aws-cdk/aws-apigateway'
import { DeployEnv } from '../config'

type Props = StackProps & {
  lambda: LambdaFunction
  resourceName: string
  deployEnv: DeployEnv
  apiName?: string
  method?: string
}

/**
 * Stack for aws-api-gateway
 */
export class APIGatewayStack extends Stack {
  public readonly restAPI: RestApi
  constructor(scope: Construct, id: string, props: Props) {
    super(scope, id, props)

    const { apiName, lambda, method, resourceName } = props

    // rest api
    this.restAPI = new RestApi(this, `ApiGateway${id}`, {
      restApiName: apiName || 'myApi',
    })

    // add Resource
    const resource = this.restAPI.root.addResource(resourceName)

    // Integration
    const lambdaIntegration = new LambdaIntegration(lambda)

    // add Method
    resource.addMethod(method || 'GET', lambdaIntegration)
  }
}
