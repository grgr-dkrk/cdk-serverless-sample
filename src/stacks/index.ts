export { LambdaStack } from './lambdaStack'
export { APIGatewayStack } from './APIGatewayStack'
