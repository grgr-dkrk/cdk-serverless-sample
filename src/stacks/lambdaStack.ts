import { Construct, Duration, Stack, StackProps } from '@aws-cdk/core'
import { Code, Function as LambdaFunction, Runtime } from '@aws-cdk/aws-lambda'
import { DeployEnv } from '../config'

type Props = StackProps & {
  code: Code
  handlerName: string
  deployEnv: DeployEnv
  functionName?: string
}

/**
 * Stack for aws-lambda
 */
export class LambdaStack extends Stack {
  public readonly lambda: LambdaFunction
  constructor(scope: Construct, id: string, props: Props) {
    super(scope, id, props)

    const { code, handlerName, functionName } = props

    this.lambda = new LambdaFunction(this, `LambdaFunc${id}`, {
      code,
      functionName: functionName || 'Lambda',
      handler: `index.${handlerName}`,
      runtime: Runtime.NODEJS_10_X,
      timeout: Duration.seconds(10),
    })
  }
}
