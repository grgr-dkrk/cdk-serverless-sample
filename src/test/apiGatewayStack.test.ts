import { Code } from '@aws-cdk/aws-lambda'
import * as cdk from '@aws-cdk/core'
import { APIGatewayStack, LambdaStack } from '../stacks'
import * as path from 'path'
import { SynthUtils } from '@aws-cdk/assert'
import { envMock } from './__mocks__/env'

describe('APIGatewayStack', () => {
  const app = new cdk.App()
  const lambdaStack = new LambdaStack(app, 'TestLambdaStack', {
    code: Code.fromAsset(path.resolve(__dirname, './__mocks__/functions')),
    deployEnv: 'staging',
    handlerName: 'LambdaFunctionMock',
    env: envMock,
  })
  const apiGatewayStack = new APIGatewayStack(app, 'TestApiGatewayStack', {
    lambda: lambdaStack.lambda,
    resourceName: 'LambdaFunctionMock',
    deployEnv: 'staging',
    env: envMock,
  })
  it('snapshot', () => {
    expect(SynthUtils.toCloudFormation(apiGatewayStack)).toMatchSnapshot(
      'apiGatewayTest',
    )
  })
})
