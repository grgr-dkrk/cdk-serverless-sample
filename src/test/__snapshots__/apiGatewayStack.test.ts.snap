// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`APIGatewayStack snapshot: apiGatewayTest 1`] = `
Object {
  "Outputs": Object {
    "ApiGatewayTestApiGatewayStackEndpointAEE08830": Object {
      "Value": Object {
        "Fn::Join": Array [
          "",
          Array [
            "https://",
            Object {
              "Ref": "ApiGatewayTestApiGatewayStack5C9F3E84",
            },
            ".execute-api.us-east-1.",
            Object {
              "Ref": "AWS::URLSuffix",
            },
            "/",
            Object {
              "Ref": "ApiGatewayTestApiGatewayStackDeploymentStageprodF1D9D672",
            },
            "/",
          ],
        ],
      },
    },
  },
  "Resources": Object {
    "ApiGatewayTestApiGatewayStack5C9F3E84": Object {
      "Properties": Object {
        "Name": "myApi",
      },
      "Type": "AWS::ApiGateway::RestApi",
    },
    "ApiGatewayTestApiGatewayStackAccount78E2AC9D": Object {
      "DependsOn": Array [
        "ApiGatewayTestApiGatewayStack5C9F3E84",
      ],
      "Properties": Object {
        "CloudWatchRoleArn": Object {
          "Fn::GetAtt": Array [
            "ApiGatewayTestApiGatewayStackCloudWatchRole5FE8AFA8",
            "Arn",
          ],
        },
      },
      "Type": "AWS::ApiGateway::Account",
    },
    "ApiGatewayTestApiGatewayStackCloudWatchRole5FE8AFA8": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "apigateway.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AmazonAPIGatewayPushToCloudWatchLogs",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "ApiGatewayTestApiGatewayStackDeployment4341927Ff42ef7ab00a3b8a6e14fff4bc7ca76fa": Object {
      "DependsOn": Array [
        "ApiGatewayTestApiGatewayStackLambdaFunctionMockGET843BC4E5",
        "ApiGatewayTestApiGatewayStackLambdaFunctionMock7B9D65BA",
      ],
      "Properties": Object {
        "Description": "Automatically created by the RestApi construct",
        "RestApiId": Object {
          "Ref": "ApiGatewayTestApiGatewayStack5C9F3E84",
        },
      },
      "Type": "AWS::ApiGateway::Deployment",
    },
    "ApiGatewayTestApiGatewayStackDeploymentStageprodF1D9D672": Object {
      "Properties": Object {
        "DeploymentId": Object {
          "Ref": "ApiGatewayTestApiGatewayStackDeployment4341927Ff42ef7ab00a3b8a6e14fff4bc7ca76fa",
        },
        "RestApiId": Object {
          "Ref": "ApiGatewayTestApiGatewayStack5C9F3E84",
        },
        "StageName": "prod",
      },
      "Type": "AWS::ApiGateway::Stage",
    },
    "ApiGatewayTestApiGatewayStackLambdaFunctionMock7B9D65BA": Object {
      "Properties": Object {
        "ParentId": Object {
          "Fn::GetAtt": Array [
            "ApiGatewayTestApiGatewayStack5C9F3E84",
            "RootResourceId",
          ],
        },
        "PathPart": "LambdaFunctionMock",
        "RestApiId": Object {
          "Ref": "ApiGatewayTestApiGatewayStack5C9F3E84",
        },
      },
      "Type": "AWS::ApiGateway::Resource",
    },
    "ApiGatewayTestApiGatewayStackLambdaFunctionMockGET843BC4E5": Object {
      "Properties": Object {
        "AuthorizationType": "NONE",
        "HttpMethod": "GET",
        "Integration": Object {
          "IntegrationHttpMethod": "POST",
          "Type": "AWS_PROXY",
          "Uri": Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":apigateway:us-east-1:lambda:path/2015-03-31/functions/",
                Object {
                  "Fn::ImportValue": "TestLambdaStack:ExportsOutputFnGetAttLambdaFuncTestLambdaStack2EE31B0BArn341416B9",
                },
                "/invocations",
              ],
            ],
          },
        },
        "ResourceId": Object {
          "Ref": "ApiGatewayTestApiGatewayStackLambdaFunctionMock7B9D65BA",
        },
        "RestApiId": Object {
          "Ref": "ApiGatewayTestApiGatewayStack5C9F3E84",
        },
      },
      "Type": "AWS::ApiGateway::Method",
    },
    "ApiGatewayTestApiGatewayStackLambdaFunctionMockGETApiPermissionTestApiGatewayStackApiGatewayTestApiGatewayStack0B074A86GETLambdaFunctionMock72246C77": Object {
      "Properties": Object {
        "Action": "lambda:InvokeFunction",
        "FunctionName": Object {
          "Fn::ImportValue": "TestLambdaStack:ExportsOutputFnGetAttLambdaFuncTestLambdaStack2EE31B0BArn341416B9",
        },
        "Principal": "apigateway.amazonaws.com",
        "SourceArn": Object {
          "Fn::Join": Array [
            "",
            Array [
              "arn:",
              Object {
                "Ref": "AWS::Partition",
              },
              ":execute-api:us-east-1:1234567890a:",
              Object {
                "Ref": "ApiGatewayTestApiGatewayStack5C9F3E84",
              },
              "/",
              Object {
                "Ref": "ApiGatewayTestApiGatewayStackDeploymentStageprodF1D9D672",
              },
              "/GET/LambdaFunctionMock",
            ],
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "ApiGatewayTestApiGatewayStackLambdaFunctionMockGETApiPermissionTestTestApiGatewayStackApiGatewayTestApiGatewayStack0B074A86GETLambdaFunctionMockF5A3DA97": Object {
      "Properties": Object {
        "Action": "lambda:InvokeFunction",
        "FunctionName": Object {
          "Fn::ImportValue": "TestLambdaStack:ExportsOutputFnGetAttLambdaFuncTestLambdaStack2EE31B0BArn341416B9",
        },
        "Principal": "apigateway.amazonaws.com",
        "SourceArn": Object {
          "Fn::Join": Array [
            "",
            Array [
              "arn:",
              Object {
                "Ref": "AWS::Partition",
              },
              ":execute-api:us-east-1:1234567890a:",
              Object {
                "Ref": "ApiGatewayTestApiGatewayStack5C9F3E84",
              },
              "/test-invoke-stage/GET/LambdaFunctionMock",
            ],
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
  },
}
`;
